### PR TITLE
Forbid AllocateValueType from allocating byref like types on the heap

### LINKED ISSUE #38769: Boxing ref struct using MethodInfo.Invoke
Hi team,

Accidentally i've discovered the ability to box a ref struct by using reflection.

The scenario is the invocation via reflection of a method that has, say, an in ref struct parameter. Upon the exit of the call the parameter array passed to MethodInfo.Invoke will contain a boxed ref struct.

Small reproduction (call Repro.Test()):
```C#
     static class Repro
    {
        ref struct RefStruct
            {
                private long L;
                private int I;
            }

        private static void MethodStr(in RefStruct rs) { }

        private static void ReflectionInvoke(string method)
        {
            var parameters = new object[1];

            var message = parameters[0] == null ? "null" : parameters[0].GetType().Name;
            Console.WriteLine(message); // Displays null.
            typeof(Repro)
                .GetMethod(method, BindingFlags.Static | BindingFlags.NonPublic)
                .Invoke(null, parameters);
            message = parameters[0] == null ? "null" : parameters[0].GetType().Name;
            Console.WriteLine(message); //Displays RefStruct.
        }

       public static void Test() => ReflectionInvoke(nameof(MethodStr));
    }
```
 
Can someone please confirm if the outcome of the second Console.WriteLine is by design?

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Resolves https://github.com/dotnet/runtime/issues/38769.

This also adds some unit test coverage gaps when calling `MethodInfo.Invoke` where some of the parameters may be byref-like structs.

### CODE REVIEW

**Comment by GrabYourPitchforks** [comment]
Need to wait for CI to build the Mono stack and see what falls out. It's possible there will be Mono-specific work as well.

**Review by davidwrighton — APPROVED** [review]

File: src/libraries/System.Runtime/tests/System/Reflection/InvokeWithRefLikeArgs.cs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+            {
+                fixed (int* pInt = &_backingField)
+                {
+                    return ref *(MyRefStruct*)pInt; // will return a valid ref
+                }
+            }
+
+            public static void TakesRefStructAsArg(MyRefStruct o)
+            {
+                Assert.Equal(0, o.MyInt); // should be default(T)
```
Comment (steveharter) [inline src/libraries/System.Runtime/tests/System/Reflection/InvokeWithRefLikeArgs.cs:108]: Do you want to set the ref struct to ensure it's not copied back?

If modified, will NotSupported be thrown, or will it just be ignored?

Comment (steveharter) [inline src/libraries/System.Runtime/tests/System/Reflection/InvokeWithRefLikeArgs.cs:108]: If you modify `o.MyInt` here (by setting the field directly) I assume that is fine -- no copy is made \ allocated.

File: src/libraries/System.Runtime/tests/System/Reflection/InvokeWithRefLikeArgs.cs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        [Fact]
+        public static void PropertyTypedAsRefToRefStruct_AsPropInfo_ThrowsNSE()
+        {
+            PropertyInfo pi = typeof(TestClass).GetProperty(nameof(TestClass.PropertyTypedAsRefToRefStruct));
+            Assert.NotNull(pi);
+            Assert.Throws<NotSupportedException>(() => pi.GetValue(null));
+        }
+
+        [Fact]
+        public static void PropertyIndexerWithRefStructArg_DoesNotCopyValueBack()
```
Comment (steveharter) [inline src/libraries/System.Runtime/tests/System/Reflection/InvokeWithRefLikeArgs.cs:78]: This worked previously, correct?

**Review by steveharter — APPROVED** [review]
A couple misc questions, but otherwise LGTM

File: src/libraries/System.Runtime/tests/System/Reflection/InvokeWithRefLikeArgs.cs (hunk 2 above)
Comment (GrabYourPitchforks) [reply src/libraries/System.Runtime/tests/System/Reflection/InvokeWithRefLikeArgs.cs:78]: Yeah. It's adding missing test coverage.

File: src/libraries/System.Runtime/tests/System/Reflection/InvokeWithRefLikeArgs.cs (hunk 1 above)
Comment (GrabYourPitchforks) [reply src/libraries/System.Runtime/tests/System/Reflection/InvokeWithRefLikeArgs.cs:108]: Correct. The struct itself is copy-by-value, and the runtime creates a new copy of it on the stack (zeroed out) before invoking the method. What this method is testing is that we didn't try doing a reinterpret_cast of unrelated memory as part of this, which could result in the data being non-zero.

**Comment by GrabYourPitchforks** [comment]
I'm going to pare the unit tests back to coreclr only and open a new tracking issue for the mono work. That way more knowledgeable folks can address it on the mono side and it won't hold up this work.

**Comment by GrabYourPitchforks** [comment]
This PR seems to have caused some test failures (see https://helix.dot.net/api/2019-06-17/jobs/8078a744-5bd3-4a70-97c9-8033e53935cb/workitems/System.Reflection.Tests/console), even though I don't think I touched any files related to those tests. Since the tests are passing on my local machine, even with a mono build, I put some debugging logic in the current PR to help narrow down the cause.

Marking no-merge while investigation takes place.

**Comment by GrabYourPitchforks** [comment]
Ah, I bet the failure is an integer overflow in these lines:

https://github.com/dotnet/runtime/blob/693cb577d08178573768d09f6961308994e3320f/src/libraries/System.Reflection/tests/TypeInfoTests.cs#L496-L497

**Pull request merged**